src/ng/directive/validators.js

```diff
@@ -162,34 +167,59 @@ var requiredDirective = function() {
  *   </file>
  * </example>
  */
-var patternDirective = function() {
+var patternDirective = function($parse) {
   return {
     restrict: 'A',
     require: '?ngModel',
-    link: function(scope, elm, attr, ctrl) {
-      if (!ctrl) return;
+    compile: function(tElm, tAttr) {
+      var patternExp;
+      var parseFn;
+
+      if (tAttr.ngPattern) {
+        patternExp = tAttr.ngPattern;
```

Comment:
Previously, `patternExp` would also be set to `attr.pattern` (if `attr.ngPattern` was undefined).
Why is this changed?

Reply:
if ngPattern isn't set, there can be no pattern expression, because pattern can only have a string value.

Reply:
Theoretically, it can have any value :grin:
(E.g. imagine a custom `myPattern` directive, that sets `$attrs.pattern`.)